enterprise/coderd/proxyhealth/proxyhealth.go

```diff
@@ -0,0 +1,235 @@
+package proxyhealth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/database/dbauthz"
+)
+
+type ProxyHealthStatus string
+
+const (
+	// Reachable means the proxy access url is reachable and returns a healthy
+	// status code.
+	Reachable ProxyHealthStatus = "reachable"
+	// Unreachable means the proxy access url is not responding.
+	Unreachable ProxyHealthStatus = "unreachable"
+	// Unregistered means the proxy has not registered a url yet. This means
+	// the proxy was created with the cli, but has not yet been started.
+	Unregistered ProxyHealthStatus = "unregistered"
+)
+
+type Options struct {
+	// Interval is the interval at which the proxy health is checked.
+	Interval   time.Duration
+	DB         database.Store
+	Logger     slog.Logger
+	Client     *http.Client
+	Prometheus *prometheus.Registry
+}
+
+// ProxyHealth runs a go routine that periodically checks the health of all
+// workspace proxies. This information is stored in memory, so each coderd
+// replica has its own view of the health of the proxies. These views should be
+// consistent, and if they are not, it indicates a problem.
+type ProxyHealth struct {
+	db       database.Store
+	interval time.Duration
+	logger   slog.Logger
+	client   *http.Client
+
+	cache *atomic.Pointer[map[uuid.UUID]ProxyStatus]
+
+	// PromMetrics
+	healthCheckDuration prometheus.Histogram
+}
+
+func New(opts *Options) (*ProxyHealth, error) {
+	if opts.Interval <= 0 {
+		opts.Interval = time.Minute
+	}
+	if opts.DB == nil {
+		return nil, xerrors.Errorf("db is required")
+	}
+	if opts.Prometheus == nil {
+		opts.Prometheus = prometheus.NewRegistry()
+	}
+
+	client := opts.Client
+	if client == nil {
+		client = http.DefaultClient
+	}
+	// Set a timeout on the client, so we don't wait forever for a healthz response.
+	tmp := *client
+	tmp.Timeout = time.Second * 5
+	client = &tmp
+
+	// Prometheus metrics
+	healthCheckDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "coderd",
+		Subsystem: "proxyhealth",
+		Name:      "health_check_duration_seconds",
+		Help:      "Histogram for duration of proxy health collection in seconds.",
+		Buckets:   []float64{0.001, 0.005, 0.010, 0.025, 0.050, 0.100, 0.500, 1, 5, 10, 30},
+	})
+	opts.Prometheus.MustRegister(healthCheckDuration)
```

Comment:
Should probably have a labeled metric too with the proxy name.

Reply:
This is the total duration to check all proxies. I just wanted to make sure this routine doesn't end up taking forever